### DISCUSSION TITLE
Feature: Readonly user

### DISCUSSION THREAD

**Discussion by shield-9** [growilabs/growi#7706] [Ideas / Requests] [answered]
登録せずにアクセスできる機能はありますが、それとは閲覧だけができるユーザが作れると便利だと思いました。

イメージとしては、既存の「管理者」「一般ユーザ」に加え、3つ目のロールとして「閲覧専用ユーザ」が選択できる感じです。

# 前提
* 外部には非公開
* どこからでもアクセスできる (接続元IPなどは限定しない)
* 利用者は不定期に増減することがある
* 編集するユーザと閲覧のみを行う利用者がはっきり分かれている

# 解決したい問題
* Basic認証+ゲストユーザのように共有の認証情報だと、定期的な変更を行わないとセキュリティが保てない
* ユーザを作成すると、本来、閲覧ができれば十分なユーザが編集もできてしまう

**Comment by yuki-takei** [comment]
@shield-9 ご提案ありがとうございます。

ざっくり見積もって、この機能はちょっとコストが高いですね…

現状[「当該ユーザーが管理者かどうか」というフラグ](https://github.com/weseek/crowi-plus/blob/master/lib/models/user.js#L47)はあるのですが、それ以外にロールという概念はなく、ゲスト閲覧可能にするオプションは

1. [オプション有効ならログインチェックを部分的にパスする](https://github.com/weseek/crowi-plus/blob/master/lib/util/middlewares.js#L204)
2. [ログインしていないユーザー向けに、ボタンを表示させなかったり disable attr を付けたりする](https://github.com/weseek/crowi-plus/blob/master/lib/views/page.html#L18)

で実現しています。2番の変更はかなり多くの箇所に及びますが、ここに「ログインはできているがロールが違う」というユーザー向けの条件を入れていくのがまず大変…それからその他の実装工数として、

* 管理ページから任意のユーザーを「閲覧専用ユーザー」に変更するという操作UI
* 「新規登録ユーザーは閲覧専用にする」というオプション

などが必要になるのではないかと。

仰るように「Basic認証+ゲストユーザ」でも一応同等のことはできてしまうので、これらを全部やって得られる利益と見合わないかなあとなんとなく感じています。

+1 リアクションがいっぱい付いたら検討するので、もし需要があるようならお友達連れてきてください笑

**Comment by yuki-takei** [comment]
#199 で Group ACL を実装中なのですが、一旦こちらで対応できるかなと思うので close します。

### ROM との細かい部分での違い

Group ACL 実装後のイメージとしては、「特定グループしか編集できないページ」を用意することで、それらのページに対しては新規ユーザーからの編集をブロックできます。

 > ユーザを作成すると、本来、閲覧ができれば十分なユーザが編集もできてしまう

と厳密には同一の機能にはなりませんが(普通にページ作ったり、全体公開ページは編集できてしまうので)、運用でカバーできる範囲かなと思います。

もっと ROM の実現に近づけるなら、admin page に Group ACL を使ったグローバル設定を設けるとかですかね。例えば、「指定グループに入っていないユーザーを ROM として扱う」とか。

**Comment by yuki-takei** [comment]
「閲覧専用ユーザ」の要望を別の方からいただいたので reopen します。
優先度は相変わらず低いままなので、実装者募集中です。

**Comment by yuki-takei** [comment, marked as answer]
起票から6年、Group ACL 実装から4年越し…

[v6.1.0](https://github.com/weseek/growi/releases/tag/v6.1.0) で実装されました。
